x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/sort/BytesRefBucketedSort.java

```diff
@@ -0,0 +1,392 @@
+        if (otherRootIndex >= other.values.size()) {
+            // The value was never collected.
+            return;
+        }
+        other.checkInvariant(bucket);
+        long otherStart = other.startIndex(otherBucket, otherRootIndex);
+        long otherEnd = other.common.endIndex(otherRootIndex);
+        // TODO: This can be improved for heapified buckets by making use of the heap structures
+        for (long i = otherStart; i < otherEnd; i++) {
+            collect(other.values.get(i).bytesRefView(), bucket);
+        }
+    }
+
+    /**
+     * Creates a block with the values from the {@code selected} groups.
+     */
+    public Block toBlock(BlockFactory blockFactory, IntVector selected) {
+        // Check if the selected groups are all empty, to avoid allocating extra memory
+        if (IntStream.range(0, selected.getPositionCount()).map(selected::getInt).noneMatch(bucket -> {
+            long rootIndex = common.rootIndex(bucket);
+            if (rootIndex >= values.size()) {
+                // Never collected
+                return false;
+            }
+            long start = startIndex(bucket, rootIndex);
+            long end = common.endIndex(rootIndex);
+            long size = end - start;
+            return size > 0;
+        })) {
+            return blockFactory.newConstantNullBlock(selected.getPositionCount());
+        }
+
+        // Used to sort the values in the bucket.
+        BytesRef[] bucketValues = new BytesRef[common.bucketSize];
+
+        try (var builder = blockFactory.newBytesRefBlockBuilder(selected.getPositionCount())) {
+            for (int s = 0; s < selected.getPositionCount(); s++) {
+                int bucket = selected.getInt(s);
+                long rootIndex = common.rootIndex(bucket);
+                if (rootIndex >= values.size()) {
+                    // Never collected
+                    builder.appendNull();
+                    continue;
+                }
+
+                long start = startIndex(bucket, rootIndex);
+                long end = common.endIndex(rootIndex);
+                long size = end - start;
+
+                if (size == 0) {
+                    builder.appendNull();
+                    continue;
+                }
+
+                if (size == 1) {
+                    try (BreakingBytesRefBuilder bytes = values.get(start)) {
+                        builder.appendBytesRef(bytes.bytesRefView());
+                    }
+                    values.set(start, null);
+                    continue;
+                }
+
+                for (int i = 0; i < size; i++) {
+                    try (BreakingBytesRefBuilder bytes = values.get(start + i)) {
+                        bucketValues[i] = bytes.bytesRefView();
+                    }
+                    values.set(start + i, null);
+                }
+
+                // TODO: Make use of heap structures to faster iterate in order instead of copying and sorting
```

Comment:
It's unrealted, but I'm thinking now: Isn't this still nlogn? Would this really be better over an in-place sort?

Saying this because we have this comment everywhere, and I'm not sure if it really can be done. Maybe I'm missing some trick

Reply:
Iterating the heap is `O(n)`, right? We aren't removing and re-heaping. We're just iterating in order.

Also it'd save a copy.

Reply:
Iterating, yes. But sorting, it's still nlogn for a heap. Unless our heapify keeps it "sortable". But I'd say that would be slower.
To sort it, the heap tells us the min value. But then, the next candidates are the 2 children. Then, it would be 3 potential candidates (1 child + 2 grand-children), and so on. Worst case, like re-heapifying on every iteration

Reply:
Sorry, yeah, it's still n log n. I presume it's better because it can rely on the heap property being there already. But I agree, it's probably not worth a ton of time on.